samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java

```diff
@@ -0,0 +1,160 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote.couchbase;
+
+import com.couchbase.client.java.Bucket;
+import com.couchbase.client.java.Cluster;
+import com.couchbase.client.java.env.CouchbaseEnvironment;
+import com.couchbase.client.java.error.TemporaryFailureException;
+import com.couchbase.client.java.error.TemporaryLockFailureException;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.io.Serializable;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.samza.context.Context;
+import org.apache.samza.operators.functions.ClosableFunction;
+import org.apache.samza.operators.functions.InitableFunction;
+import org.apache.samza.serializers.Serde;
+
+
+public abstract class BaseCouchbaseTableFunction<V> implements InitableFunction, ClosableFunction, Serializable {
+
+  // Clients
+  private final static CouchbaseBucketRegistry COUCHBASE_BUCKET_REGISTRY = new CouchbaseBucketRegistry();
+  protected transient CouchbaseEnvironment env;
+  protected transient Cluster cluster;
+  protected transient Bucket bucket;
+
+  // Function Settings
+  protected Serde<V> valueSerde = null;
+  protected Long timeout = 0L;
+  protected TimeUnit timeUnit = null;
+  protected Integer ttl = 0; // default value 0 means no ttl, data will be stored forever
+
+  // Cluster Settings
+  protected final List<String> clusterNodes;
+  protected final String bucketName;
+
+  // Environment Settings
+  protected CouchbaseEnvironmentConfigs environmentConfigs;
+
+  public BaseCouchbaseTableFunction(String bucketName, List<String> clusterNodes, Class<V> valueClass) {
+    Preconditions.checkArgument(StringUtils.isNotEmpty(bucketName), "Bucket name is not allowed to be null or empty.");
+    Preconditions.checkArgument(CollectionUtils.isNotEmpty(clusterNodes),
+        "Cluster nodes is not allowed to be null or empty.");
+    Preconditions.checkArgument(valueClass != null, "Value class is not allowed to be null.");
+    this.bucketName = bucketName;
+    this.clusterNodes = ImmutableList.copyOf(clusterNodes);
+    this.environmentConfigs = new CouchbaseEnvironmentConfigs();
+  }
+
+  @Override
+  public void init(Context context) {
+    bucket = COUCHBASE_BUCKET_REGISTRY.getBucket(bucketName, clusterNodes, environmentConfigs);
+  }
+
+  @Override
+  public void close() {
```

Comment:
Can we not use the same approach as in LI version, and handle close centrally. So we don't need the reference counting stuff?

Reply:
The way samza-li is using to close the buckets is that, after application has shutdown, offspring will ask the bucketCreator to close all the opened buckets.
However, in open source, do we still have any control to the buckets after the application has shutdown?